src/main/java/org/springframework/integration/kafka/listener/WindowingOffsetManager.java

```diff
@@ -0,0 +1,310 @@
+
+	private volatile boolean closed;
+
+	public WindowingOffsetManager(OffsetManager offsetManager) {
+		this.delegate = offsetManager;
+	}
+
+	/**
+	 * The timespan for aggregating write operations, before invoking the underlying {@link OffsetManager}.
+	 * @param timespan duration in milliseconds
+	 */
+	public void setTimespan(long timespan) {
+		Assert.isTrue(timespan >= 0, "Timespan must be a positive value");
+		this.timespan = timespan;
+	}
+
+	/**
+	 * How many writes should be aggregated, before invoking the underlying {@link OffsetManager}. Setting this value
+	 * to 1 effectively disables windowing.
+	 * @param count number of writes
+	 */
+	public void setCount(int count) {
+		Assert.isTrue(count >= 0, "Count must be a positive value");
+		this.count = count;
+	}
+
+	/**
+	 * The timeout that {@link #close()} and {@link #destroy()}
+	 * operations will wait for receiving a confirmation that
+	 * the underlying writes have been processed.
+	 * @param shutdownTimeout duration in milliseconds
+	 */
+	public void setShutdownTimeout(int shutdownTimeout) {
+		this.shutdownTimeout = shutdownTimeout;
+	}
+
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		if (this.count != 1) {
+			createOffsetsStream();
+		}
+	}
+
+	private void createOffsetsStream() {
+		if (!this.closed) {
+			this.offsetsLock.writeLock().lock();
+			try {
+				this.offsets = RingBufferProcessor.share("spring-integration-kafka-offset", 1024);
+			}
+			finally {
+				this.offsetsLock.writeLock().unlock();
+			}
+			Streams.wrap(this.offsets)
+					.window(this.count, timespan, TimeUnit.MILLISECONDS)
+					.flatMap(findHighestOffsetsByPartition)
+					.consume(this.delegateUpdateOffset, null, this.offsetComplete);
+
+		}
+	}
+
+	@Override
+	public void destroy() throws Exception {
+		flush();
+		close();
+		if (this.delegate instanceof DisposableBean) {
+			((DisposableBean) this.delegate).destroy();
+		}
+	}
+
+	@Override
+	public void updateOffset(Partition partition, long offset) {
+		if (this.offsets != null) {
+			this.offsetsLock.readLock().lock();
+			try {
+				this.offsets.onNext(new PartitionAndOffset(partition, offset));
+			}
+			finally {
+				this.offsetsLock.readLock().unlock();
+			}
+		}
+		else {
+			this.delegate.updateOffset(partition, offset);
+		}
+	}
+
+	@Override
+	public long getOffset(Partition partition) {
+		doFlush();
+		return this.delegate.getOffset(partition);
+	}
+
+	@Override
+	public void deleteOffset(Partition partition) {
+		doFlush();
+		this.delegate.deleteOffset(partition);
+	}
+
+	@Override
+	public void resetOffsets(Collection<Partition> partition) {
+		doFlush();
+		this.delegate.resetOffsets(partition);
+	}
+
+	@Override
+	public void close() throws IOException {
+		this.closed = true;
+		this.delegate.close();
+	}
+
+	@Override
+	public void flush() throws IOException {
+		if (this.offsets != null) {
+			this.offsets.awaitAndShutdown(this.shutdownTimeout, TimeUnit.MILLISECONDS);
```

Comment:
Haven't actually tried it, but doesn't this call prevent the `RingBufferProcessor` to be used afterwards? Essentially any call to `delete` or `reset` preventing any update afterwards?